src/replays.ts

```diff
@@ -0,0 +1,299 @@
+					query.append(`ORDER BY ${order} DESC)`);
+					query.append(SQL` ORDER BY ${order} DESC LIMIT ${limit1}, 51;`);
+					return replays.query(query, db);
+				} else {
+					const query = SQL`(SELECT uploadtime, id, format, p1, p2, password FROM ps_replays `;
+					query.append(`FORCE INDEX (p1) `);
+					query.append(SQL`WHERE private = ${isPrivate} AND p1id = ${userid} ORDER BY ${order} DESC)`);
+					query.append(' UNION ');
+					query.append('(SELECT uploadtime, id, format, p1, p2, password FROM ps_replays FORCE INDEX (p2) ');
+					query.append(SQL`WHERE private = ${isPrivate} AND p2id = ${userid} ORDER BY ${order} DESC)`);
+					query.append(SQL` ORDER BY ${order} DESC LIMIT ${limit1}, 51;`);
+					return replays.query(query, db);
+				}
+			}
+		}
+
+		if (args.byRating) {
+			const query = SQL`SELECT uploadtime, id, format, p1, p2, rating, password `;
+			query.append(`FROM ps_replays FORCE INDEX (top) `);
+			query.append(SQL`WHERE private = ${isPrivate} AND formatid = ${format} ORDER BY rating DESC LIMIT 51`);
+			return replays.query(query, db);
+		} else {
+			const query = SQL`SELECT uploadtime, id, format, p1, p2, rating, password `;
+			query.append(`FROM ps_replays FORCE INDEX (format) `);
+			query.append(SQL`WHERE private = ${isPrivate} AND formatid = ${format} ORDER BY rating DESC LIMIT 51`);
+			return replays.query(query, db);
+		}
+	}
+
+	async fullSearch(term: string, page = 0) {
+		if (page > 0) return [];
+
+		const patterns = term.split(',').map(subterm => {
+			const escaped = subterm.replace(/%/g, '\\%').replace(/_/g, '\\_');
+			return `%${escaped}%`;
+		});
+		if (patterns.length !== 1 && patterns.length !== 2) return [];
+
+		const query = SQL`SELECT /*+ MAX_EXECUTION_TIME(10000) */ `;
+		query.append(`uploadtime, id, format, p1, p2, password FROM ps_replays `);
+		query.append(SQL`FORCE INDEX (recent) WHERE private = 0 AND log LIKE ${patterns[0]} `);
+		if (patterns.length === 2) {
+			query.append(SQL`AND log LIKE ${patterns[1]} `);
+		}
+		query.append(`ORDER BY uploadtime DESC LIMIT 10;`);
+
+		return replays.query(query, db);
+	}
+
+	async recent() {
+		const query = SQL`SELECT uploadtime, id, format, p1, p2 FROM ps_replays `;
+		query.append(`FORCE INDEX (recent) WHERE private = 0 ORDER BY uploadtime `);
+		query.append('DESC LIMIT 50');
+		return replays.query(query, db);
+	}
+
+	normalizeUsername(username: string) {
+		return username.toLowerCase().replace(/[^A-Za-z0-9]+/g, '');
+	}
+
+	async upload(params: {[k: string]: any}, dispatcher: Dispatcher) {
+		let id = params.id;
+		if (!toID(id)) throw new ActionError('Battle ID needed.');
+		const preppedReplay = await prepreplays.get('*', id, db);
+		const replay = await replays.get(['id', 'private', 'password'], id, db);
+		if (!preppedReplay) {
+			if (replay) {
+				if (replay.password) {
+					id += '-' + replay.password + 'pw';
+				}
+				return 'success:' + id;
```

Comment:
What is with the result values here? Is this part of the PS protocol? Generally you'd want something more typed like `{type: "success", id}`

Reply:
Yeah, this is part of the protocol. I don't like it either, but it's not something to change in this migration. After the fact, sure.

Reply:
That's unfortunate. I guess that's fine but I'll see if I can think of a better way to do this internally